common/dynamicconfig/collection.go

```diff
@@ -92,6 +92,9 @@ const (
 	errCountLogThreshold = 1000
 )
 
+// ErrInvalidConfiguration indicates that the value provided by dynamic config is not legal
+var ErrInvalidConfiguration = errors.New("invalid dynamic configuration")
```

Comment:
I've given this some thought. The ErrInvalidConfiguration error generated by getConfiguredWorkersCountFor(namespace) wrapper function. The functions only responsibility is to validate the value retrieved from dynamic config—whether it's from the configuration file or defaults if the value can't be read.

To keep things simple and easy to handle down the road, it makes sense to put all the rules for checking in one place. The natural spot for these rules is in the config reader. This way, we get to choose: do we want to give a thumbs-up during system startup if the config isn't playing by the rules, or just roll with the default values?

Considering this rationale, the checks outlined in getConfiguredWorkersCountFor need to be executed as part of the matchAndConvert process within the dynamicconfig package. Consequently, I've designated the error as a generic error within the dynamicconfig package.

Reply:
More validation within dynamicconfig may be a good thing, but that's not happening in this PR. So this error shouldn't be here. If this package does more validation, then it can have more errors.

Note that dynamic config usage is pretty free-form, so adding validation won't be easy. Clients can ask for any values at any time, and each request might come with a different default value (and validation options if added). Keys are not all known statically. Plus, we have to keep the interface backwards-compatible to avoid breaking code that embeds the server.

(Also I think you mean heads-up, not thumbs-up :) )

Reply:
Please remove this error